src/storage/src/sst/stream_writer.rs

```diff
@@ -0,0 +1,196 @@
+            buffer,
+            bytes_written: Default::default(),
+            flushed: false,
+            threshold: buffer_threshold,
+            arrow_schema: arrow_schema.clone(),
+        })
+    }
+
+    /// Write a record batch to stream writer.
+    pub async fn write(&mut self, batch: &Batch) -> error::Result<()> {
+        let arrow_batch = RecordBatch::try_new(
+            self.arrow_schema.clone(),
+            batch
+                .columns()
+                .iter()
+                .map(|v| v.to_arrow_array())
+                .collect::<Vec<_>>(),
+        )
+        .context(NewRecordBatchSnafu)?;
+        self.arrow_writer
+            .write(&arrow_batch)
+            .context(WriteParquetSnafu)?;
+        let written = Self::try_flush(
+            &self.path,
+            &self.buffer,
+            &mut self.object_writer,
+            false,
+            &mut self.flushed,
+            self.threshold,
+        )
+        .await?;
+        self.bytes_written.fetch_add(written, Ordering::Relaxed);
+        Ok(())
+    }
+
+    /// Abort writer.
+    pub async fn abort(self) -> bool {
+        // TODO(hl): Currently we can do nothing if file's parts have been uploaded to remote storage
+        // on abortion, we need to find a way to abort the upload. see https://help.aliyun.com/document_detail/31996.htm?spm=a2c4g.11186623.0.0.3eb42cb7b2mwUz#reference-txp-bvx-wdb
+        !self.flushed
+    }
+
+    /// Close parquet writer and ensure all buffered data are written into underlying storage.
+    pub async fn close(mut self) -> error::Result<(FileMetaData, u64)> {
+        let metadata = self.arrow_writer.close().context(WriteParquetSnafu)?;
+        let written = Self::try_flush(
+            &self.path,
+            &self.buffer,
+            &mut self.object_writer,
+            true,
+            &mut self.flushed,
+            self.threshold,
+        )
+        .await?;
+        self.bytes_written.fetch_add(written, Ordering::Relaxed);
+        self.object_writer
+            .close()
+            .await
+            .context(WriteObjectSnafu { path: &self.path })?;
+        Ok((metadata, self.bytes_written.load(Ordering::Relaxed)))
+    }
+
+    /// Try to flush buffered data to underlying storage if it's size exceeds threshold.
+    /// Set `all` to true if all buffered data should be flushed regardless of it's size.
+    async fn try_flush(
+        file_name: &str,
+        shared_buffer: &Buffer,
+        object_writer: &mut Writer,
+        all: bool,
+        flushed: &mut bool,
+        threshold: usize,
+    ) -> error::Result<u64> {
+        let mut bytes_written = 0;
+
+        // Once buffered data size reaches threshold, split the data in chunks (typically 4MB)
+        // and write to underlying storage.
+        while shared_buffer.buffer.lock().unwrap().len() >= threshold {
+            let chunk = {
+                let mut buffer = shared_buffer.buffer.lock().unwrap();
```

Comment:
nit perf: maybe these two lock acquiring actions can be merged?

Reply:
> nit perf: maybe these two lock acquiring actions can be merged?

`shared_buffer.buffer.lock().unwrap().len()` and `let mut buffer = shared_buffer.buffer.lock().unwrap();` cannot be merged into one lock acquisition, otherwise the `MutexGuard` will be held across await point, causing the async function returns a future that is not `Send` because `MutexGuard` is `!Send`